File: CrossNavigation/InteractiveTransition/CNInteractiveTransition.m
//
// CNInteractiveTransition.m
//
// Copyright (c) 2014 Artem Stepanenko
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "CNInteractiveTransition.h"
#import "CNViewController.h"
#import "CNTransitioningFactory.h"
#import "CNTransitioning.h"
#import "cn_utils.h"

@interface CNInteractiveTransition ()

@property (nonatomic, weak) CNViewController *fromViewController;
@property (nonatomic, weak) CNViewController *toViewController;

@property (nonatomic, strong) CNTransitioning *appearanceTransitioning;
@property (nonatomic, strong) CNTransitioning *disappearanceTransitioning;

@property (nonatomic, strong) id<UIViewControllerContextTransitioning> transitionContext;
@property (nonatomic, strong) UIView *containerView;

@end

@implementation CNInteractiveTransition

- (instancetype)init
{
    self = [super init];
    
    if (self) {
        _direction = CNDirectionNone;
    }
    
    return self;
}

- (void)setDirection:(CNDirection)direction
{
    _direction = direction;
    
    self.appearanceTransitioning = [CNTransitioningFactory transitioningForDirection:direction present:YES];
    self.disappearanceTransitioning = [CNTransitioningFactory transitioningForDirection:direction present:NO];
}

#pragma mark - UIViewControllerTransitioningDelegate

- (id<UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented
                                                                  presentingController:(UIViewController *)presenting
                                                                      sourceController:(UIViewController *)source
{
    return self.appearanceTransitioning;
}

- (id<UIViewControllerAnimatedTransitioning>)animationControllerForDismissedController:(UIViewController *)dismissed
{
    return self.disappearanceTransitioning;
}

- (id<UIViewControllerInteractiveTransitioning>)interactionControllerForPresentation:(id <UIViewControllerAnimatedTransitioning>)animator
{
    if (self.interactive) {
        return self;
    }
    
    return nil;
}

- (id<UIViewControllerInteractiveTransitioning>)interactionControllerForDismissal:(id<UIViewControllerAnimatedTransitioning>)animator
{
    if (self.interactive) {
        return self;
    }
    
    return nil;
}

#pragma mark - Interactive Transitioning

- (void)startInteractiveTransition:(id<UIViewControllerContextTransitioning>)transitionContext
{
    self.transitionContext = transitionContext;
    self.containerView = [self.transitionContext containerView];
    
    CGRect containerFrame = self.containerView.bounds;
    
    // from view controller
    self.fromViewController = (CNViewController *)[self.transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
    self.fromViewController.view.frame = [self.appearanceTransitioning fromViewStartFrameWithContainerFrame:containerFrame];
    [self.containerView addSubview:self.fromViewController.view];
    
    // to view controller
    self.toViewController = (CNViewController *)[self.transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
    self.toViewController.view.frame = [self.appearanceTransitioning toViewStartFrameWithContainerFrame:containerFrame];
    [self.containerView addSubview:self.toViewController.view];
}

- (void)updateInteractiveTransition:(CGFloat)percentComplete
{
    CGRect containerFrame = self.containerView.bounds;
    
    // normalization
    percentComplete = MAX(0.0f, percentComplete);
    percentComplete = MIN(1.0f, percentComplete);
    
    // update fromViewController's frame
    self.fromViewController.view.frame = CGRectTransform([self.appearanceTransitioning fromViewStartFrameWithContainerFrame:containerFrame],
                                                         [self.appearanceTransitioning fromViewEndFrameWithContainerFrame:containerFrame],
                                                         percentComplete);
    
    // update toViewController's frame
    self.toViewController.view.frame = CGRectTransform([self.appearanceTransitioning toViewStartFrameWithContainerFrame:containerFrame],
                                                       [self.appearanceTransitioning toViewEndFrameWithContainerFrame:containerFrame],
                                                       percentComplete);
    
}

- (void)finishInteractiveTransition
{
    [self finishInteractiveTransition:YES];
}

- (void)cancelInteractiveTransition
{
    [self finishInteractiveTransition:NO];
}

#pragma mark - Private

- (void)finishInteractiveTransition:(BOOL)didComplete
{
    CGRect containerFrame = self.containerView.bounds;
    CGAffineTransform transform = self.containerView.transform;
    
    [UIView animateWithDuration:[CNTransitioning duration] animations:^{
        
        if (didComplete) {
            self.fromViewController.view.frame = [self.appearanceTransitioning fromViewEndFrameWithContainerFrame:containerFrame];
            self.toViewController.view.frame = [self.appearanceTransitioning toViewEndFrameWithContainerFrame:containerFrame];
        } else {
            self.fromViewController.view.frame = [self.appearanceTransitioning fromViewStartFrameWithContainerFrame:containerFrame];
            self.toViewController.view.frame = [self.appearanceTransitioning toViewStartFrameWithContainerFrame:containerFrame];
        }
        
    } completion:^(BOOL finished) {
        
        [self.transitionContext completeTransition:didComplete];
        self.transitionContext = nil;
        self.containerView = nil;
        
        // reason of existence of these lines is Apple's bug
        if (didComplete) {
            self.toViewController.view.transform = transform;
            self.toViewController.view.frame = self.toViewController.view.superview.bounds;
        } else {
            self.fromViewController.view.transform = transform;
            self.fromViewController.view.frame = self.fromViewController.view.superview.bounds;
        }
    }];
}

@end

File: Swift/CrossNavigation/InteractiveTransition/Transioning/CNInteractiveTransition.swift
//
//  CNInteractiveTransition.swift
//  CrossNavigationDemo
//
//  Created by Nicolas Purita on 11/5/14.
//
//

import UIKit

class CNInteractiveTransition: UIPercentDrivenInteractiveTransition, UIViewControllerTransitioningDelegate {
   
    // private vars
    private var appearanceTransitioning: CNTransitioning?
    private var disappearanceTransitioning: CNTransitioning?
    private var transitionContext: UIViewControllerContextTransitioning?
    private var interactive: Bool = true

    // read-only
    private(set) var fromViewController: CNViewController?
    private(set) var toViewController: CNViewController?
    private(set) var containerView: UIView?
    private(set) var recentPercentComplete: CGFloat = 0

    // public var
    var finishingDuration: CGFloat = CGFloat(CNTransitioning().duration)
    var direction: CNDirection {
        didSet {
            self.appearanceTransitioning = CNTranstioningFactory.Transitioning(direction, present: true)
            self.disappearanceTransitioning = CNTranstioningFactory.Transitioning(direction, present: false)
        }
    }
    
    override init() {
        self.direction = .None
    }
    
    // MARK: Interactive Transtioning
    
    override func startInteractiveTransition(transitionContext: UIViewControllerContextTransitioning) -> () {
        self.transitionContext = transitionContext
        self.containerView = self.transitionContext!.containerView()
        
        let containerFrame = self.containerView!.bounds
        
        // from View Controller
        self.fromViewController = self.transitionContext!.viewControllerForKey(UITransitionContextFromViewControllerKey) as? CNViewController
        self.fromViewController!.view.frame = self.appearanceTransitioning!.fromViewStartFrameWithContainerFrame(containerFrame)
        self.containerView!.addSubview(self.fromViewController!.view)
        
        // to View Controller
        self.toViewController = self.transitionContext!.viewControllerForKey(UITransitionContextToViewControllerKey) as? CNViewController
        self.toViewController!.view.frame = self.appearanceTransitioning!.toViewStartFrameWithContainerFrame(containerFrame)
        self.containerView!.addSubview(self.toViewController!.view)
    }
    
    func updateinteractiveTransition(percentComplete: CGFloat) -> () {
        let containerFrame = self.containerView!.bounds
        
        // normalization
        var mutatedPercentComplete = percentComplete
        mutatedPercentComplete = max(0.0, mutatedPercentComplete)
        mutatedPercentComplete = min(1.0, mutatedPercentComplete)
        
        self.recentPercentComplete = mutatedPercentComplete
        
        // update fromViewController's frame
        self.fromViewController!.view.frame = CGRect.Transform(self.appearanceTransitioning!.fromViewStartFrameWithContainerFrame(containerFrame),
            finalRect: self.appearanceTransitioning!.fromViewEndFrameWithContainerFrame(containerFrame), rate: self.recentPercentComplete)
        
        // update toViewController's frame
        self.toViewController!.view.frame = CGRect.Transform(self.appearanceTransitioning!.toViewStartFrameWithContainerFrame(containerFrame),
            finalRect: self.appearanceTransitioning!.toViewEndFrameWithContainerFrame(containerFrame), rate: self.recentPercentComplete)
    }
    
    func finishInteractiveTranstion() -> () {
        self.finishInteractiveTransition(true)
    }
    
    func cancelInteractiveTranstion() -> ()  {
        self.finishInteractiveTransition(false)
    }
    
    // MARK: Private methods
    
    private func finishInteractiveTransition(didComplete: Bool) -> () {

        let containerFrame = self.containerView!.bounds
        let transform = self.containerView!.transform
        let duration: NSTimeInterval = NSTimeInterval(self.finishingDuration)
        
        UIView.animateWithDuration(duration, animations: { () -> Void in
            
            if didComplete {
                self.fromViewController!.view.frame = self.appearanceTransitioning!.fromViewEndFrameWithContainerFrame(containerFrame)
                self.toViewController!.view.frame = self.appearanceTransitioning!.toViewEndFrameWithContainerFrame(containerFrame)
            } else {
                self.fromViewController!.view.frame = self.appearanceTransitioning!.fromViewStartFrameWithContainerFrame(containerFrame)
                self.toViewController!.view.frame = self.appearanceTransitioning!.toViewStartFrameWithContainerFrame(containerFrame)
            }
            
        }) { (completion: Bool) -> Void in
            
            self.transitionContext!.completeTransition(didComplete)
            self.transitionContext = nil
            self.containerView = nil
            
            if didComplete {
                self.toViewController!.view.transform = transform
                self.toViewController!.view.frame = self.toViewController!.view.superview!.bounds
            } else {
                self.fromViewController!.view.transform = transform
                self.fromViewController!.view.frame = self.fromViewController!.view.superview!.bounds
            }
            
        }
    }

    // MARK: UIViewControllerTransitioningDelegate
    
    func animationControllerForPresentedController(presented: UIViewController, presentingController presenting: UIViewController, sourceController source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        return self.appearanceTransitioning
    }
    
    func animationControllerForDismissedController(dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        return self.disappearanceTransitioning
    }
    
    func interactionControllerForPresentation(animator: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        return self.interactive ? self : nil
    }
    
    func interactionControllerForDismissal(animator: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        return self.interactive ? self : nil
    }
    
}
